src/H5Tconv.c

```diff
@@ -2753,131 +2795,201 @@ H5T__conv_struct_opt(H5T_t *src, H5T_t *dst, H5T_cdata_t *cdata, const H5T_conv_
  */
 static herr_t
-H5T__conv_enum_init(H5T_t *src, H5T_t *dst, H5T_cdata_t *cdata)
-{
-    H5T_enum_struct_t *priv = NULL;         /*private conversion data    */
-    int                n;                   /*src value cast as native int    */
-    int                domain[2] = {0, 0};  /*min and max source values    */
-    int               *map       = NULL;    /*map from src value to dst idx    */
-    unsigned           length;              /*nelmts in map array        */
-    unsigned           i, j;                /*counters            */
-    herr_t             ret_value = SUCCEED; /* Return value */
+H5T__conv_enum_init(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata, const H5T_conv_ctx_t *conv_ctx)
+{
+    H5T_conv_enum_t *priv          = NULL; /* Private conversion data */
+    int             *map           = NULL; /* Map from src value to dst idx */
+    bool             rebuild_cache = false;
+    herr_t           ret_value     = SUCCEED;
 
     FUNC_ENTER_PACKAGE
 
     cdata->need_bkg = H5T_BKG_NO;
-    if (NULL == (priv = (H5T_enum_struct_t *)(cdata->priv = H5MM_calloc(sizeof(*priv)))))
-        HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "memory allocation failed");
-    if (0 == src->shared->u.enumer.nmembs)
-        HGOTO_DONE(SUCCEED);
 
-    /*
-     * Check that the source symbol names are a subset of the destination
-     * symbol names and build a map from source member index to destination
-     * member index.
-     */
-    H5T__sort_name(src, NULL);
-    H5T__sort_name(dst, NULL);
-    if (NULL == (priv->src2dst = (int *)H5MM_malloc(src->shared->u.enumer.nmembs * sizeof(int))))
-        HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "memory allocation failed");
-    for (i = 0, j = 0; i < src->shared->u.enumer.nmembs && j < dst->shared->u.enumer.nmembs; i++, j++) {
-        while (j < dst->shared->u.enumer.nmembs &&
-               strcmp(src->shared->u.enumer.name[i], dst->shared->u.enumer.name[j]) != 0)
-            j++;
-        if (j >= dst->shared->u.enumer.nmembs)
-            HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
-                        "source type is not a subset of destination type");
-        priv->src2dst[i] = (int)j;
-    } /* end for */
+    priv = (H5T_conv_enum_t *)(cdata->priv);
+    if (!priv) {
+        if (NULL == (priv = (H5T_conv_enum_t *)(cdata->priv = calloc(1, sizeof(*priv)))))
+            HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "memory allocation failed");
+        rebuild_cache = true;
+    }
+    else {
+        /* Check if we need to rebuild our cache. For now, treat
+         * enums as different even if one is just a subset of the
+         * other
+         */
+        if (cdata->command == H5T_CONV_CONV && conv_ctx->u.conv.recursive)
+            /* Recursive conversion; we can reuse the cache */
+            rebuild_cache = false;
+        else {
+            if (0 != H5T_cmp(src, priv->src_copy, false) || 0 != H5T_cmp(dst, priv->dst_copy, false))
```

Comment:
How does this not cause problems if there's an enum inside a compound? Wouldn't the compound code mark it as recursive then this enum code would fail to calculate the mapping?

Reply:
At this point, the compound init should have called `H5T_path_find` for all the member types, initing the enum path in the process and building/rebuilding the cache as necessary. At the point we get here from conversion command with this marked as recursive, we should have a valid cache that we can just use for each of the compound elements being converted rather than checking if we need to rebuild the cache for every element.